crates/wit-parser/src/ast/resolve.rs

```diff
@@ -31,22 +35,65 @@ enum Key {
 }
 
 impl Resolver {
-    pub(crate) fn resolve(&mut self, ast: &ast::Ast<'_>) -> Result<Document> {
-        let mut interface_map = IndexMap::new();
-
+    fn resolve_interfaces(&mut self, ast: &ast::Ast<'_>) -> Result<()> {
+        let mut deps: HashMap<&str, HashSet<&str>> = HashMap::new();
+        // Register all interface names
         for interface in ast.interfaces() {
-            let name = &interface.name.name;
-            let instance = self.resolve_interface(name, &interface.items, &interface.docs)?;
-
-            if interface_map.insert(name.to_string(), instance).is_some() {
+            let name = &*interface.name.name;
+            if deps.insert(name, HashSet::new()).is_some() {
                 return Err(Error {
                     span: interface.name.span,
                     msg: format!("interface `{name}` is defined more than once"),
                 }
-                .into());
+                .into())
+            }
+        }
+        // ...then build dependency graph
+        for interface in ast.interfaces() {
+            let name = &*interface.name.name;
+            for use_ in interface.uses() {
+                let use_name = &*use_.from.name;
+                if !deps.contains_key(use_name) {
+                    return Err(Error {
+                        span: use_.from.span,
+                        msg: format!("interface `{use_name}` is undefined"),
+                    }
+                    .into())
+                }
+                deps
+                    .get_mut(name)
+                    .unwrap()
+                    .insert(use_name);
             }
         }
 
+        let toposort = TopoSort::from_map(deps).into_vec_nodes();
+        let sorted = match toposort {
+            SortResults::Full(names) => {
+                names
+                    .into_iter()
+                    .map(|name| ast
+                        .interfaces()
+                        .find(|i| &*i.name.name == name)
+                        .unwrap())
+            },
+            SortResults::Partial(_) => {
+                bail!("interface `use` cycle detected");
+            },
```

Comment:
I think it's definitely ok to punt "really good error messages" to later, but I do think this should at least mention one interface (with a span) in the chain of cyclic interfaces

Reply:
How does this look? The following `wit` ...

```
interface x {
    type m = string
    use { m } from y
}

interface y {
    use { m } from x
}
```

... produces the error:
```
Error: interface y contains a `use` cycle
     --> test.wit:6:11
      |
    6 | interface y {
      |
```